File: backend/minigames/powerups.py
from __future__ import annotations

import random
from typing import Any

import numpy as np

from Config import SingletonConfig

from .session import MinigameSessionState


POWERUP_KEYS = ("bomb", "glove", "twist")


def _default_count_for_difficulty(difficulty: int) -> int:
    return 1 if int(difficulty) == 1 else 5


def _new_game_default_counts(state: MinigameSessionState) -> dict[str, int]:
    if state.engine is not None and getattr(state.engine, "legacy_name", "") == "Blitzkrieg":
        return {"bomb": 10, "glove": 10, "twist": 10}
    default_count = _default_count_for_difficulty(state.difficulty)
    return {key: default_count for key in POWERUP_KEYS}


def load_powerup_counts(state: MinigameSessionState) -> dict[str, int]:
    if state.engine is None:
        return {"bomb": 0, "glove": 0, "twist": 0}
    config = SingletonConfig().config
    difficulty_state = config.setdefault("power_ups_state", [dict(), dict()])[state.difficulty]
    saved_counts = difficulty_state.get(getattr(state.engine, "legacy_name", ""), None)
    if not saved_counts:
        return _new_game_default_counts(state)
    values = list(saved_counts)
    while len(values) < len(POWERUP_KEYS):
        values.append(_default_count_for_difficulty(state.difficulty))
    return {key: max(0, int(values[index])) for index, key in enumerate(POWERUP_KEYS)}


def save_powerup_counts(state: MinigameSessionState) -> None:
    if state.engine is None:
        return
    config = SingletonConfig().config
    difficulty_state = config.setdefault("power_ups_state", [dict(), dict()])[state.difficulty]
    difficulty_state[getattr(state.engine, "legacy_name", "")] = tuple(
        int(state.powerup_counts.get(key, 0)) for key in POWERUP_KEYS
    )
    SingletonConfig().save_config(config)


def reset_powerup_counts(state: MinigameSessionState) -> None:
    state.powerup_counts = _new_game_default_counts(state)
    save_powerup_counts(state)


def _valid_bomb_targets(engine) -> list[int]:
    return [index for index, value in enumerate(engine.board.flatten().tolist()) if int(value) > 0]


def _valid_glove_sources(engine) -> list[int]:
    return [index for index, value in enumerate(engine.board.flatten().tolist()) if int(value) > 0]


def _valid_glove_targets(engine) -> list[int]:
    return [index for index, value in enumerate(engine.board.flatten().tolist()) if int(value) == 0]


def _valid_twist_targets(engine) -> list[int]:
    valid_targets: list[int] = []
    for row in range(engine.rows - 1):
        for col in range(engine.cols - 1):
            sub_board = np.array(engine.board[row : row + 2, col : col + 2], copy=False)
            if np.all(sub_board == 0):
                continue
            if np.any(sub_board == -1):
                continue
            valid_targets.append(row * engine.cols + col)
    return valid_targets


def build_powerups_payload(state: MinigameSessionState) -> dict[str, Any]:
    enabled = state.engine is not None
    return {
        "enabled": enabled,
        "counts": {key: int(state.powerup_counts.get(key, 0)) for key in POWERUP_KEYS},
        "activeMode": state.active_mode,
    }


def build_interaction_payload(state: MinigameSessionState) -> dict[str, Any]:
    engine = state.engine
    if engine is None or not state.active_mode:
        return {
            "active": False,
            "mode": None,
            "targetType": None,
            "phase": None,
            "validTargets": [],
            "selectedIndices": [],
            "hintKey": "",
        }

    if state.active_mode == "bomb":
        return {
            "active": True,
            "mode": "bomb",
            "targetType": "singleTile",
            "phase": "target",
            "validTargets": _valid_bomb_targets(engine),
            "selectedIndices": [],
            "hintKey": "minigames.powerups.hints.bomb",
        }

    if state.active_mode == "glove" and state.interaction_phase == 1:
        return {
            "active": True,
            "mode": "glove",
            "targetType": "singleTile",
            "phase": "source",
            "validTargets": _valid_glove_sources(engine),
            "selectedIndices": [],
            "hintKey": "minigames.powerups.hints.gloveSource",
        }

    if state.active_mode == "glove" and state.interaction_phase == 2:
        source_index = int((state.selection_cache or {}).get("source_index", -1))
        return {
            "active": True,
            "mode": "glove",
            "targetType": "singleTile",
            "phase": "target",
            "validTargets": _valid_glove_targets(engine),
            "selectedIndices": [source_index] if source_index >= 0 else [],
            "hintKey": "minigames.powerups.hints.gloveTarget",
        }

    if state.active_mode == "twist":
        return {
            "active": True,
            "mode": "twist",
            "targetType": "subgridTopLeft",
            "phase": "target",
            "validTargets": _valid_twist_targets(engine),
            "selectedIndices": [],
            "hintKey": "minigames.powerups.hints.twist",
        }

    return {
        "active": False,
        "mode": None,
        "targetType": None,
        "phase": None,
        "validTargets": [],
        "selectedIndices": [],
        "hintKey": "",
    }


def cancel_powerup_interaction(
    state: MinigameSessionState, *, clear_animation: bool = True
) -> None:
    if clear_animation and state.engine is not None:
        state.engine.clear_animation()
    state.active_mode = None
    state.interaction_phase = 0
    state.selection_cache = None


def activate_powerup(state: MinigameSessionState, mode: str) -> bool:
    engine = state.engine
    normalized_mode = str(mode or "").strip().lower()
    if engine is None or normalized_mode not in POWERUP_KEYS:
        return False
    engine.clear_animation()
    if int(state.powerup_counts.get(normalized_mode, 0)) <= 0:
        return False
    if state.active_mode == normalized_mode:
        cancel_powerup_interaction(state)
        return True
    state.active_mode = normalized_mode
    state.interaction_phase = 1
    state.selection_cache = None
    return True


def _consume_powerup(state: MinigameSessionState, mode: str) -> None:
    state.powerup_counts[mode] = max(0, int(state.powerup_counts.get(mode, 0)) - 1)
    save_powerup_counts(state)


def _post_powerup_update(state: MinigameSessionState) -> None:
    engine = state.engine
    if engine is None:
        return
    engine.check_game_passed()
    engine.check_game_over()
    engine.save_to_config()


def _apply_bomb(state: MinigameSessionState, index: int) -> bool:
    engine = state.engine
    if engine is None:
        return False
    row, col = divmod(index, engine.cols)
    if row < 0 or row >= engine.rows or col < 0 or col >= engine.cols:
        engine.clear_animation()
        return False
    if int(engine.board[row, col]) <= 0:
        engine.clear_animation()
        return False
    engine.board[row, col] = 0
    engine.set_special_effects([{"type": "explosion", "index": int(index)}])
    _consume_powerup(state, "bomb")
    _post_powerup_update(state)
    return True


def _apply_glove_step(state: MinigameSessionState, index: int) -> bool:
    engine = state.engine
    if engine is None:
        return False
    row, col = divmod(index, engine.cols)
    if row < 0 or row >= engine.rows or col < 0 or col >= engine.cols:
        cancel_powerup_interaction(state)
        return False

    if state.interaction_phase == 1:
        value = int(engine.board[row, col])
        if value <= 0:
            engine.clear_animation()
            return False
        state.interaction_phase = 2
        state.selection_cache = {"source_index": int(index), "value": value}
        engine.set_special_effects(
            [{"type": "grab", "index": int(index), "value": int(value)}]
        )
        return False

    source_index = int((state.selection_cache or {}).get("source_index", -1))
    if source_index < 0:
        cancel_powerup_interaction(state)
        return False

    source_row, source_col = divmod(source_index, engine.cols)
    source_value = int(engine.board[source_row, source_col])
    success = int(engine.board[row, col]) == 0 and source_value > 0
    if success:
        engine.board[row, col] = source_value
        engine.board[source_row, source_col] = 0
        engine.set_special_effects(
            [
                {
                    "type": "glove_move",
                    "fromIndex": int(source_index),
                    "toIndex": int(index),
                    "value": int(source_value),
                }
            ]
        )
        _consume_powerup(state, "glove")
        _post_powerup_update(state)
    else:
        engine.clear_animation()
    cancel_powerup_interaction(state, clear_animation=not success)
    return success


def _rotate_2x2_clockwise(board: np.ndarray, row: int, col: int) -> None:
    sub_board = np.array(board[row : row + 2, col : col + 2], copy=True)
    board[row : row + 2, col : col + 2] = np.array(
        [[sub_board[1, 0], sub_board[0, 0]], [sub_board[1, 1], sub_board[0, 1]]],
        dtype=board.dtype,
    )


def _apply_twist(state: MinigameSessionState, index: int) -> bool:
    engine = state.engine
    if engine is None:
        return False
    row, col = divmod(index, engine.cols)
    if row < 0 or col < 0 or row >= engine.rows - 1 or col >= engine.cols - 1:
        engine.clear_animation()
        return False
    sub_board = np.array(engine.board[row : row + 2, col : col + 2], copy=False)
    if np.all(sub_board == 0) or np.any(sub_board == -1):
        engine.clear_animation()
        return False
    top_left = row * engine.cols + col
    top_right = top_left + 1
    bottom_left = (row + 1) * engine.cols + col
    bottom_right = bottom_left + 1
    twist_tiles = []
    twist_mapping = (
        (top_left, top_right, int(engine.board[row, col])),
        (top_right, bottom_right, int(engine.board[row, col + 1])),
        (bottom_left, top_left, int(engine.board[row + 1, col])),
        (bottom_right, bottom_left, int(engine.board[row + 1, col + 1])),
    )
    for from_index, to_index, value in twist_mapping:
        if value > 0:
            twist_tiles.append(
                {
                    "fromIndex": int(from_index),
                    "toIndex": int(to_index),
                    "value": int(value),
                }
            )
    _rotate_2x2_clockwise(engine.board, row, col)
    engine.set_special_effects(
        [
            {
                "type": "twist",
                "index": int(index),
                "tiles": twist_tiles,
            }
        ]
    )
    _consume_powerup(state, "twist")
    _post_powerup_update(state)
    return True


def apply_target_action(state: MinigameSessionState, index: int) -> bool:
    if state.engine is None or not state.active_mode:
        return False
    success = False
    if state.active_mode == "bomb":
        success = _apply_bomb(state, index)
        cancel_powerup_interaction(state, clear_animation=not success)
        return success
    if state.active_mode == "glove":
        return _apply_glove_step(state, index)
    if state.active_mode == "twist":
        success = _apply_twist(state, index)
        cancel_powerup_interaction(state, clear_animation=not success)
        return success
    cancel_powerup_interaction(state)
    return False


def maybe_award_random_powerup(state: MinigameSessionState, score_delta: int) -> str | None:
    if score_delta < 300 or score_delta >= 1024:
        return None
    probability = 0.05 if int(state.difficulty) == 1 else 0.25
    if random.random() >= probability:
        return None
    awarded = random.choice(list(POWERUP_KEYS))
    state.powerup_counts[awarded] = int(state.powerup_counts.get(awarded, 0)) + 1
    save_powerup_counts(state)
    return awarded

File: frontend/src/features/minigames/engine/powerups.js
import { flattenBoard, POWERUP_KEYS } from './utils.js';

export function defaultPowerupCounts(state) {
  if (state.engine?.legacyName === 'Blitzkrieg') {
    return { bomb: 10, glove: 10, twist: 10 };
  }
  const count = Number(state.difficulty) === 1 ? 1 : 5;
  return { bomb: count, glove: count, twist: count };
}

export function buildPowerupsPayload(state) {
  return {
    enabled: Boolean(state.engine),
    counts: Object.fromEntries(POWERUP_KEYS.map((key) => [key, Number(state.powerupCounts?.[key] || 0)])),
    activeMode: state.activeMode || null,
  };
}

function validBombTargets(engine) {
  return flattenBoard(engine.board).reduce((targets, value, index) => {
    if (value > 0) targets.push(index);
    return targets;
  }, []);
}

function validGloveSources(engine) {
  return validBombTargets(engine);
}

function validGloveTargets(engine) {
  return flattenBoard(engine.board).reduce((targets, value, index) => {
    if (value === 0) targets.push(index);
    return targets;
  }, []);
}

function validTwistTargets(engine) {
  const targets = [];
  for (let row = 0; row < engine.rows - 1; row += 1) {
    for (let col = 0; col < engine.cols - 1; col += 1) {
      const values = [
        engine.board[row][col],
        engine.board[row][col + 1],
        engine.board[row + 1][col],
        engine.board[row + 1][col + 1],
      ];
      if (values.every((value) => value === 0)) continue;
      if (values.some((value) => value === -1)) continue;
      targets.push(row * engine.cols + col);
    }
  }
  return targets;
}

export function buildInteractionPayload(state) {
  const engine = state.engine;
  if (!engine || !state.activeMode) {
    return {
      active: false,
      mode: null,
      targetType: null,
      phase: null,
      validTargets: [],
      selectedIndices: [],
      hintKey: '',
    };
  }

  if (state.activeMode === 'bomb') {
    return {
      active: true,
      mode: 'bomb',
      targetType: 'singleTile',
      phase: 'target',
      validTargets: validBombTargets(engine),
      selectedIndices: [],
      hintKey: 'minigames.powerups.hints.bomb',
    };
  }

  if (state.activeMode === 'glove' && state.interactionPhase === 1) {
    return {
      active: true,
      mode: 'glove',
      targetType: 'singleTile',
      phase: 'source',
      validTargets: validGloveSources(engine),
      selectedIndices: [],
      hintKey: 'minigames.powerups.hints.gloveSource',
    };
  }

  if (state.activeMode === 'glove' && state.interactionPhase === 2) {
    const sourceIndex = Number(state.selectionCache?.sourceIndex ?? -1);
    return {
      active: true,
      mode: 'glove',
      targetType: 'singleTile',
      phase: 'target',
      validTargets: validGloveTargets(engine),
      selectedIndices: sourceIndex >= 0 ? [sourceIndex] : [],
      hintKey: 'minigames.powerups.hints.gloveTarget',
    };
  }

  if (state.activeMode === 'twist') {
    return {
      active: true,
      mode: 'twist',
      targetType: 'subgridTopLeft',
      phase: 'target',
      validTargets: validTwistTargets(engine),
      selectedIndices: [],
      hintKey: 'minigames.powerups.hints.twist',
    };
  }

  return {
    active: false,
    mode: null,
    targetType: null,
    phase: null,
    validTargets: [],
    selectedIndices: [],
    hintKey: '',
  };
}

export function cancelPowerupInteraction(state, { clearAnimation = true } = {}) {
  if (clearAnimation && state.engine) {
    state.engine.clearAnimation();
  }
  state.activeMode = null;
  state.interactionPhase = 0;
  state.selectionCache = null;
}

export function activatePowerup(state, mode) {
  const normalizedMode = String(mode || '').trim().toLowerCase();
  if (!state.engine || !POWERUP_KEYS.includes(normalizedMode)) {
    return false;
  }
  state.engine.clearAnimation();
  if (Number(state.powerupCounts?.[normalizedMode] || 0) <= 0) {
    return false;
  }
  if (state.activeMode === normalizedMode) {
    cancelPowerupInteraction(state);
    return true;
  }
  state.activeMode = normalizedMode;
  state.interactionPhase = 1;
  state.selectionCache = null;
  return true;
}

function consumePowerup(state, mode) {
  state.powerupCounts[mode] = Math.max(0, Number(state.powerupCounts?.[mode] || 0) - 1);
}

function postPowerupUpdate(state) {
  const engine = state.engine;
  if (!engine) return;
  engine.checkGamePassed();
  engine.checkGameOver();
}

function applyBomb(state, index) {
  const engine = state.engine;
  if (!engine) return false;
  const row = Math.floor(index / engine.cols);
  const col = index % engine.cols;
  if (row < 0 || row >= engine.rows || col < 0 || col >= engine.cols) {
    engine.clearAnimation();
    return false;
  }
  if (Number(engine.board[row][col]) <= 0) {
    engine.clearAnimation();
    return false;
  }
  engine.board[row][col] = 0;
  engine.setSpecialEffects([{ type: 'explosion', index: Number(index) }]);
  consumePowerup(state, 'bomb');
  postPowerupUpdate(state);
  return true;
}

function applyGloveStep(state, index) {
  const engine = state.engine;
  if (!engine) return false;
  const row = Math.floor(index / engine.cols);
  const col = index % engine.cols;
  if (row < 0 || row >= engine.rows || col < 0 || col >= engine.cols) {
    cancelPowerupInteraction(state);
    return false;
  }

  if (state.interactionPhase === 1) {
    const value = Number(engine.board[row][col]) || 0;
    if (value <= 0) {
      engine.clearAnimation();
      return false;
    }
    state.interactionPhase = 2;
    state.selectionCache = { sourceIndex: Number(index), value };
    engine.setSpecialEffects([{ type: 'grab', index: Number(index), value }]);
    return false;
  }

  const sourceIndex = Number(state.selectionCache?.sourceIndex ?? -1);
  if (sourceIndex < 0) {
    cancelPowerupInteraction(state);
    return false;
  }
  const sourceRow = Math.floor(sourceIndex / engine.cols);
  const sourceCol = sourceIndex % engine.cols;
  const sourceValue = Number(engine.board[sourceRow][sourceCol]) || 0;
  const success = Number(engine.board[row][col]) === 0 && sourceValue > 0;
  if (success) {
    engine.board[row][col] = sourceValue;
    engine.board[sourceRow][sourceCol] = 0;
    engine.setSpecialEffects([
      {
        type: 'glove_move',
        fromIndex: sourceIndex,
        toIndex: Number(index),
        value: sourceValue,
      },
    ]);
    consumePowerup(state, 'glove');
    postPowerupUpdate(state);
  } else {
    engine.clearAnimation();
  }
  cancelPowerupInteraction(state, { clearAnimation: !success });
  return success;
}

function rotate2x2Clockwise(board, row, col) {
  const topLeft = board[row][col];
  const topRight = board[row][col + 1];
  const bottomLeft = board[row + 1][col];
  const bottomRight = board[row + 1][col + 1];
  board[row][col] = bottomLeft;
  board[row][col + 1] = topLeft;
  board[row + 1][col] = bottomRight;
  board[row + 1][col + 1] = topRight;
}

function applyTwist(state, index) {
  const engine = state.engine;
  if (!engine) return false;
  const row = Math.floor(index / engine.cols);
  const col = index % engine.cols;
  if (row < 0 || col < 0 || row >= engine.rows - 1 || col >= engine.cols - 1) {
    engine.clearAnimation();
    return false;
  }
  const values = [
    engine.board[row][col],
    engine.board[row][col + 1],
    engine.board[row + 1][col],
    engine.board[row + 1][col + 1],
  ];
  if (values.every((value) => value === 0) || values.some((value) => value === -1)) {
    engine.clearAnimation();
    return false;
  }
  const topLeft = row * engine.cols + col;
  const topRight = topLeft + 1;
  const bottomLeft = (row + 1) * engine.cols + col;
  const bottomRight = bottomLeft + 1;
  const mapping = [
    [topLeft, topRight, engine.board[row][col]],
    [topRight, bottomRight, engine.board[row][col + 1]],
    [bottomLeft, topLeft, engine.board[row + 1][col]],
    [bottomRight, bottomLeft, engine.board[row + 1][col + 1]],
  ];
  const twistTiles = mapping
    .filter(([_from, _to, value]) => Number(value) > 0)
    .map(([fromIndex, toIndex, value]) => ({ fromIndex, toIndex, value: Number(value) }));
  rotate2x2Clockwise(engine.board, row, col);
  engine.setSpecialEffects([{ type: 'twist', index: Number(index), tiles: twistTiles }]);
  consumePowerup(state, 'twist');
  postPowerupUpdate(state);
  return true;
}

export function applyTargetAction(state, index) {
  if (!state.engine || !state.activeMode) {
    return false;
  }
  let success = false;
  if (state.activeMode === 'bomb') {
    success = applyBomb(state, Number(index));
    cancelPowerupInteraction(state, { clearAnimation: !success });
    return success;
  }
  if (state.activeMode === 'glove') {
    return applyGloveStep(state, Number(index));
  }
  if (state.activeMode === 'twist') {
    success = applyTwist(state, Number(index));
    cancelPowerupInteraction(state, { clearAnimation: !success });
    return success;
  }
  cancelPowerupInteraction(state);
  return false;
}

export function maybeAwardRandomPowerup(state, scoreDelta) {
  if (Number(scoreDelta) < 300 || Number(scoreDelta) >= 1024) {
    return null;
  }
  const probability = Number(state.difficulty) === 1 ? 0.05 : 0.25;
  if (Math.random() >= probability) {
    return null;
  }
  const awarded = POWERUP_KEYS[Math.floor(Math.random() * POWERUP_KEYS.length)];
  state.powerupCounts[awarded] = Number(state.powerupCounts?.[awarded] || 0) + 1;
  return awarded;
}

export function clonePowerupCounts(counts, fallback) {
  return Object.fromEntries(
    POWERUP_KEYS.map((key) => [key, Math.max(0, Number(counts?.[key] ?? fallback?.[key] ?? 0) || 0)])
  );
}
